src/Scope.spec.ts

```diff
@@ -484,6 +484,148 @@ describe('Scope', () => {
+                    buz as myOtherClass
+                end class
+            `);
+            await program.validate();
+            expect(program.getDiagnostics().map(x => x.message)).to.eql([
+                DiagnosticMessages.expectedValidTypeToFollowAsKeyword().message,
+                DiagnosticMessages.expectedValidTypeToFollowAsKeyword().message
+            ]);
+        });
+
+        it('finds custom types inside namespaces', async () => {
+            await program.addOrReplaceFile({ src: s`${rootDir}/source/main.bs`, dest: s`source/main.bs` }, `
+                namespace MyNamespace
+                    class MyClass
+                    end class
+
+                    function foo(param as MyClass) as MyClass
+                    end function
+
+                    function bar(param as MyNamespace.MyClass) as MyNamespace.MyClass
+                    end function
+
+                end namespace
+
+            `);
+            await program.validate();
+
+            expect(program.getDiagnostics()[0]?.message).not.to.exist;
+        });
+
+        it('finds custom types from other namespaces', async () => {
+            await program.addOrReplaceFile({ src: s`${rootDir}/source/main.bs`, dest: s`source/main.bs` }, `
+                namespace MyNamespace
+                    class MyClass
+                    end class
+                end namespace
+
+                function foo(param as MyNamespace.MyClass) as MyNamespace.MyClass
+                end function
+            `);
+            await program.validate();
+
+            expect(program.getDiagnostics()[0]?.message).not.to.exist;
+        });
+
+        it('detects missing custom types from current namespaces', async () => {
+            await program.addOrReplaceFile({ src: s`${rootDir}/source/main.bs`, dest: s`source/main.bs` }, `
+                namespace MyNamespace
+                    class MyClass
+                    end class
+
+                    function foo() as UnknownType
+                    end function
+                end namespace
+            `);
+            await program.validate();
+
+            expect(program.getDiagnostics().map(x => x.message)).to.eql([
+                DiagnosticMessages.invalidFunctionReturnType('UnknownType').message
+            ]);
+        });
+
+        it('detects missing custom types from another namespaces', async () => {
+            await program.addOrReplaceFile({ src: s`${rootDir}/source/main.bs`, dest: s`source/main.bs` }, `
+                namespace MyNamespace
+                    class MyClass
+                    end class
+                end namespace
+
+                function foo() as MyNamespace.UnknownType
+                end function
+            `);
+            await program.validate();
+
+            expect(program.getDiagnostics().map(x => x.message)).to.eql([
+                DiagnosticMessages.invalidFunctionReturnType('MyNamespace.UnknownType').message
+            ]);
+        });
+
```

Comment:
Could you add a few tests where the custom type exists in a different file than the function referencing it? Like a test where it finds the custom type in a different file, and another test where it doesn't include the custom type because it's in a separate scope (you'll need to add an xml file with the script reference for this one).

Reply:
Also, a test where the custom type is defined in the parent xml component, and referenced in the child xml component.